src/plugins/intel_cpu/src/nodes/tensoriterator.cpp

```diff
@@ -568,6 +579,41 @@ void TensorIterator::prepareParams() {
 void TensorIterator::execute(dnnl::stream strm) {
     //Special case, the subgraph is dynamic while the node has all static shapes
     if (runAsDynamic()) {
+        // redefine memory for input nodes of subgraph and
+        // reset first_mappers as the primitives are invalid.
+        auto restoreSubgraphInputByBackEdges = [&]() {
+            for (auto input_map : backedge_input_mappers) {
+                const auto extern_input_index = std::get<0>(input_map);
+                const auto body_input_index = std::get<1>(input_map);
+                auto from_mem = getParentEdgesAtPort(extern_input_index)[0]->getMemoryPtr();
+                auto new_dims = from_mem->getShape();
+                auto &to_mems = input_mems[body_input_index];
+                const auto& body_inshape = to_mems.front()->getShape();
+                OPENVINO_ASSERT(!body_inshape.isDynamic());     // Should not be dynamic as the body inputs have already been reshaped.
+                if (body_inshape != new_dims) {
+                    const auto desc = std::make_shared<CpuBlockedMemoryDesc>(to_mems.front()->getDesc().getPrecision(), Shape(new_dims));
+                    redefineToMemories(to_mems, desc);
+
+                    // update first_mappers to replace its legacy input memory addr.
+                    bool success = false;   // there must be one and only one first_mapper item corresponding to a back_edge.
+                    for (auto& first_mapper : first_mappers) {
+                        auto mapper = std::dynamic_pointer_cast<BackEdgePortHelper>(first_mapper);
+                        OPENVINO_ASSERT(mapper);
+                        auto map_rule = mapper->getMapRule();
```

Comment:
PortMapHelper objects loses PortMap info, thus we need a method to identify which first_mapper is impacted by back edge. Its mem_holder_dst is invalidated by back edge in this situation.
@maxnick

Reply:
applied review comments.